riscv/src/register/mcause.rs

```diff
@@ -51,68 +56,52 @@ pub enum Exception {
     InstructionPageFault = 12,
     LoadPageFault = 13,
     StorePageFault = 15,
-    Unknown,
 }
 
-impl From<usize> for Interrupt {
+/// SAFETY: `Exception` represents the standard RISC-V exceptions
+unsafe impl ExceptionNumber for Exception {
+    const MAX_EXCEPTION_NUMBER: usize = Self::StorePageFault as usize;
+
     #[inline]
-    fn from(nr: usize) -> Self {
-        match nr {
-            1 => Self::SupervisorSoft,
-            3 => Self::MachineSoft,
-            5 => Self::SupervisorTimer,
-            7 => Self::MachineTimer,
-            9 => Self::SupervisorExternal,
-            11 => Self::MachineExternal,
-            _ => Self::Unknown,
-        }
+    fn number(self) -> usize {
+        self as usize
     }
-}
-
-impl TryFrom<Interrupt> for usize {
-    type Error = Interrupt;
 
     #[inline]
-    fn try_from(value: Interrupt) -> Result<Self, Self::Error> {
-        match value {
-            Interrupt::Unknown => Err(Self::Error::Unknown),
-            _ => Ok(value as Self),
+    fn from_number(value: usize) -> Result<Self, usize> {
+        if value == 10 || value == 14 || value > 15 {
+            Err(value)
+        } else {
+            // SAFETY: valid exception number
+            unsafe { Ok(core::mem::transmute::<usize, Self>(value)) }
```

Comment:
Same as above, try replacing with a `match` on expected values.

Reply:
Same here, but with three comparisons

Reply:
> Same here, but with three comparisons

This is incorrect, the only valid exceptions are `0..=15`. This implementation allows any number that isn't `10`, `14`, or `15`.

What happens when someone calls:

```rust
Exception::from_number(0xbad)
```

Reply:
> Same here, but with three comparisons

This is incorrect, the only valid `Exception` variants are `0..=15`. This implementation allows any number that isn't `10`, `14`, or `15`.

What happens when someone calls:

```rust
Exception::from_number(0xbad)
```

Reply:
Same here, the output of three `cargo bench` runs:

```
    Running benches/match.rs (target/release/deps/match-570076475774c958)
Gnuplot not found, using plotters backend
from_match              time:   [114.02 ps 114.13 ps 114.23 ps]
                        change: [-25.459% -24.130% -23.185%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) low severe
  7 (7.00%) low mild

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [110.27 ps 110.48 ps 110.72 ps]
                        change: [-29.280% -27.967% -27.074%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
```

---

```
     Running benches/match.rs (target/release/deps/match-570076475774c958)
Gnuplot not found, using plotters backend
from_match              time:   [113.96 ps 114.14 ps 114.30 ps]
                        change: [-0.6957% -0.2491% +0.2621%] (p = 0.32 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [110.88 ps 111.15 ps 111.37 ps]
                        change: [-0.3696% +0.0428% +0.5089%] (p = 0.85 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) low mild
```

---

```
test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches/match.rs (target/release/deps/match-570076475774c958)
Gnuplot not found, using plotters backend
from_match              time:   [109.67 ps 109.85 ps 110.05 ps]
                        change: [-3.1358% -2.7367% -2.3282%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [113.24 ps 113.52 ps 113.81 ps]
                        change: [+2.0011% +2.4700% +2.8800%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
```

Reply:
Same here, the output of three `cargo bench` runs:

```
     Running benches/match.rs (target/release/deps/match-570076475774c958)
Gnuplot not found, using plotters backend
from_match              time:   [112.89 ps 113.23 ps 113.57 ps]
                        change: [+1.7402% +2.1406% +2.5643%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [111.94 ps 112.12 ps 112.31 ps]
                        change: [-1.3954% -0.8420% -0.3073%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) low severe
  3 (3.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe
```

---

```
     Running benches/match.rs (target/release/deps/match-570076475774c958)
Gnuplot not found, using plotters backend
from_match              time:   [111.00 ps 111.12 ps 111.24 ps]
                        change: [-1.6529% -1.3156% -0.9716%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low severe
  4 (4.00%) low mild
  1 (1.00%) high severe

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [111.62 ps 111.81 ps 112.00 ps]
                        change: [-0.6062% -0.0364% +0.5361%] (p = 0.90 > 0.05)
                        No change in performance detected.
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  3 (3.00%) high mild
  1 (1.00%) high severe
```

---

```
Gnuplot not found, using plotters backend
from_match              time:   [113.22 ps 113.63 ps 113.99 ps]
                        change: [+0.1077% +0.8447% +1.5072%] (p = 0.02 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low severe
  1 (1.00%) high mild

     Running benches/transmute.rs (target/release/deps/transmute-b6412ad997b8e934)
Gnuplot not found, using plotters backend
from_transmute          time:   [114.66 ps 115.01 ps 115.39 ps]
                        change: [+1.6948% +2.2350% +2.7761%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low severe
  3 (3.00%) low mild
```

With 16 variants:

```rust
#[repr(usize)]
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum F {
    A = 1,
    B = 2,
    C = 3,
    D = 4,
    E = 5,
    F = 6,
    G = 7,
    H = 8,
    I = 9,
    J = 10,
    K = 11,
    L = 12,
    M = 13,
    N = 14,
    O = 15,
    P = 16,
}

impl F {
    pub fn from_num_transmute(val: usize) -> Option<Self> {
        if val == 0 || val > 16 {
            None
        } else {
            Some(unsafe { std::mem::transmute::<usize, Self>(val) })
        }
    }

    pub fn from_num_match(val: usize) -> Option<Self> {
        match val {
            1 => Some(Self::A),
            2 => Some(Self::B),
            3 => Some(Self::C),
            4 => Some(Self::D),
            5 => Some(Self::E),
            6 => Some(Self::F),
            7 => Some(Self::G),
            8 => Some(Self::H),
            9 => Some(Self::I),
            10 => Some(Self::J),
            11 => Some(Self::K),
            12 => Some(Self::L),
            13 => Some(Self::M),
            14 => Some(Self::N),
            15 => Some(Self::O),
            16 => Some(Self::P),
            _ => None,
        }
    }
}
```

Reply:
Soooo... looks like LLVM is smarter than I thought (again). I will move to the classic, safe match fashion ASAP :) Thank you so much!